### DISCUSSION TITLE
이벤트 리스너 내에서의 this는 어떻게 될까요?

### DISCUSSION THREAD

**Discussion by hwangstar156** [woowacourse-study/2022-book-muscle#23] [온스타] [answered]
```jsx
class Onstar {
  constructor() {
    this.$clickBox = document.querySelector('#click-box');
  }

  bindEvent() {
    this.$clickBox.addEventListener('click', this.clickEvent);
  }

  clickEvent() {
    console.log(this);
  }
}

const onstar = new Onstar();
onstar.bindEvent();
```
여기서 과연 클릭했을때 this에서는 뭐가 나올까요? class의 메서드니까 그 상위요소인 `Onstar 인스턴스` ? 
만약 그렇게 나온다면 저희는 bind(this) 이런것을 하지않아도 바로 될텐데 우리는 왜 .bind(this)를 이용할까요?

**Comment by onschan** [comment, marked as answer]
정리된 글을 읽으면 이 내용이 궁금했을게 당연합니다!! 바로 다음 내용이 콜백함수에 대한 내용이었거든요! ㅋㅋㅋ
다음 글을 좀 스포하면서, 자스민이 질문해주었던 this, bind를 콜백함수의 관점에서 얘기를 해보자면 다음과 같습니다!

함수A의 제어권을 다른 함수B(또는 메서드) 에게 넘겨주는 경우에, 함수A를 `콜백 함수`라고 합니다.
이때 함수A의 this 는 함수B의 내부 로직에서 정한 규칙에 따라 값이 결정되는데요! (함수 B에 따라 다르다~)

자스민이 질문으로 올려주신 콜백 함수를 인자로 받는 addEventListener 메서드는 내부로직에 따라 콜백 함수(핸들러)를 호출 할때 자신의 this 를 상속하도록 정의되어있습니다.

읽어보셨던 글에 따른다면 addEventListener의 this가 뭔지는 쉽게 알 수 있겠죠?
`.` 앞의 `this.$clickBox` 즉, `document.querySelector('#click-box')라는 엘리먼트`가 addEventListener가 가리키는 this 가 되겠네요!

그러면 자스민이 올려주신 코드에서 저 'click-box' 라는 이름을 가진 박스를 클릭하면 출력되는 값 또한 알 수 있겠네요!
콜백 함수로 호출된 clickEvent 에서의 this는 addEventListener의 로직에 따라 addEventListener가 가리키는 this인 `document.querySelector('#click-box')를 통해 얻어온 엘리먼트` 를 가리킬 것이고, 
따라서 `document.querySelector('#click-box')를 통해 얻어온 엘리먼트` 가 출력되겠네요.
음.. 말이 길어져서 예시로 들자면 `<span id="click-box">클릭</span>` 이런식의 결과가 출력 될 겁니다.

그러면 자스민이 원했던 것 처럼 this가 Onstar를 가리키고있지않네요! 
지금 상황에선 this 바인딩을 따로 안하고 해당 this로 사용해도 문제가 없겠지만 
```js
class Onstar {
  constructor() {
    this.$clickBox = document.querySelector('#click-box');
  }

  bindEvent() {
    this.$clickBox.addEventListener('click', this.clickEvent);
  }

  clickEvent() {
    this.sayHello();
  }

  sayHello() {
    console.log("Hello");
  }
}

const onstar = new Onstar();
onstar.bindEvent();
```
이렇게 핸들러 내부에서 this를 이용하여 다른 메서드를 불러온 경우는 어떻게 될까요?
clickEvent 핸들러가 호출되고 동작할때 JS 엔진은 이렇게 동작할겁니다.
```
clickEvent 호출
this.sayHello() 발견
> this 가 가리키는 대상이 뭐지?
> id가 click-box인 엘리먼트를 가리키고있네?
> 뭐야 이 엘리먼트에는 sayHello 라는 메서드가 없는데?
> 에러 출력!! "this.sayHell is not a Function"
```

이때 우리는 JS 엔진이 이 this를 엘리먼트가 아닌 Onstar 로 바라봐주길 원할겁니다!
그래야 내부의 메서드에 접근할 수 있으니깐요!

이미 잘 아시겠지만 bind(this) 는 이 때 아주 유용하게 동작합니다.  한번 같이 살펴보죠!

```js
class Onstar {
  ...
  bindEvent() {
    this.$clickBox.addEventListener('click', this.clickEvent.bind(this));  
  }
  ...
}
```

이렇게 콜백함수 뒤에 .bind(this)를 붙여볼까요? 
이때 bind는 자신이 가져온 매개변수 this로 clickEvent의 this를 바꿔줍니다.
bind가 호출될 당시 this 는 Onstar 를 가리키고있겠죠?  클래스 내부에서 사용하는 this는 해당 클래스를 가리키니깐요!

즉, 콜백함수로 호출된 clickEvent 내부에서 클릭할 박스 엘리먼트를 가리키고있던 this가 
bind의 this인 Onstar로 변경되겠네요!

그러면 아까 JS 엔진이 this를 엘리먼트로 보고 에러를 출력했던 부분에서
this 를 Onstar로 인식하고 메서드를 호출하겠네요!

```
clickEvent 호출
this.sayHello() 발견
> this 가 가리키는 대상이 뭐지?
> Onstar를 가리키고있네?
> Onstar를 살펴보니 메서드로 sayHello를 가지고있네
> "Hello" 출력!
```

정리하자면 콜백 함수를 받는 메서드의 내부 동작 방식에 따라 this 가 가리키는 대상이 달라 질 수가 있고 (바인딩되는 this가 궁금하면 찾아봐야겠네요..!)
자스민이 원한 것처럼 this가 상위요소를 가리키고 싶다면 this가 상위요소를 가리키는 상황에서 bind 메서드를 이용할 수 있다!
이렇게 이해하면 될 것같아요!

---
### 더 알아보기
 
- 우리가 흔히 사용하는 setTimeout 함수와 forEach 메서드 등은 콜백함수의 this를 바인딩하는 내부로직이
addEventListenr 와 다르게 this가 전역객체를 가리키도록 되어있습니다.
- bind 의 동작방식은 this 가 가리키는 대상이 다른 새로운 함수를 생성하는 것입니다.

---
콜백함수가 가리키는 this 의 특징과, 그에 대한 의문과 해결에 대해서 알아봤습니다!
혹시 이해가 안된다거나 제가 이해하고 설명한 부분 중 잘못된게 있다면 저한테도 알려주세요!!

**Reply by hwangstar156** [reply]
이정도로 자세하게 답변이 올줄 몰랐네요! addEventListener가 자신의 this를 callback에게 상속하고 bind가 인수로 전해진 this로 callback에 들어갈 this를 변환시켜주는군요. 사실 결과는 알고있었지만 이런 자세한 동작방식은 처음알게됬네요. 고마워요 온스타❤️❤️

**Comment by wonsss** [comment]
그리고 콜백함수 호출 시 bind(this)로 해당 인스턴스를 바인딩을 하는 방법 외에도, 화살표 함수를 사용하는 방법도 있는 것 같아요.

```jsx
// 화살표 함수 사용 방법
// 매개변수로 `e`를 넣었는데, 해당 이벤트의 정보를 받는 상황을 전제했습니다.
class Onstar {
  ...
  bindEvent() {
    this.$clickBox.addEventListener('click', (e) => this.clickEvent(e));  
  }

  clickEvent(e) {
		const somethingToUse = e.target.value;
	}
  ...
}
``` 
다시 bind 사용 방법으로 돌아와서, 추가적으로 살펴보겠습니다.
```jsx
// bind 사용 방법
class Onstar {
  ...
	// 신기하게도 this로 바인딩을 하면, event에서 발생한 객체를 매개변수로 명시하지 않아도 알아서 넘어간다. 
  // 오히려 아래 콜백함수 호출 시 두번째 매개변수에 event를 추가하면 오류가 발생합니다.
  bindEvent() {
    this.$clickBox.addEventListener('click', this.clickEvent.bind(this));  
  }

  clickEvent(e) {
		const somethingToUse = e.target.value;
	}
  ...
}
``` 
이 외에도 binding 방법에는 call도 있는데요! bind는 함수가 가리키는 this는 바꾸면서 새로운 함수를 생성하는데, 생성된 함수가 호출되지는 않아요. 이벤트가 발생해야만 호출이 되겠죠. 
call은 다음과 같은 코드가 동일한 효과를 발생시켜요.
```jsx
// call 사용 방법
class Onstar {
  ...
	// call로 바인딩을 할 때, 콜백함수가 event를 매개변수로 필요하다면 두 번째 인자로 event 객체도 넣어줘야 합니다 
  bindEvent() {
    this.$clickBox.addEventListener('click', (e) => this.clickEvent.call(this, e));  
  }

  clickEvent(e) {
		const somethingToUse = e.target.value;
	}
  ...
}
```

아직 bind, call의 정확한 차이를 이해하지 못해서 잘 설명하긴 어려우나,
코드는 위와 같이 작동하는 것 같습니다. 

작동원리에 대해 상세히 설명해주실 수 있는 분 계실까요?

**Reply by onschan** [reply]
댓글 감사합니다 마르코~ 콜백함수의 다양한 방식에 대한 좋은 접근이네요👍!
저도 bind 메서드에 대한 이해가 없었을땐 콜백함수 넘길때 습관적으로 마르코가 처음 말한 방식인 
```js 
bindEvent() {
    this.$clickBox.addEventListener('click', (e) => this.clickEvent(e));  
  }
````
이런 식으로 넘겼었어요! 
그 이유는 제가 원하는 인자를 콜백함수에 넘겨주기 편하다는 이유였는데 생각해보면 bind 메서드의 첫번째 이후의 인자들로 넘겨줘도 되긴하겠네요!

이미 너무 잘 알고계신거같아서 작동원리에 대해 설명할게 없는거같은데...

콜백함수에 event 를 넘겨주지 않아도 되는 것은 addEventListener 메서드가 콜백함수를 다루는 과정에서 자동적으로 인자로 넣어주는게 아닌가 싶어요! 
화살표 함수의 첫번째 인자로 들어가는 e가 적용되는걸 보면 addEventListener 내부에서 콜백함수의 첫번째 인자로 해당 엘리먼트의 이벤트를 부여하고 해당 이벤트를 사용한다는 것을 어느정도 확인할 수 있을거같은데,

bind 메서드 사용에서는 함수 실행이 아닌 새로 생성된 함수 자체가 콜백함수로 넘어갔기 때문에 이벤트 인자가 자동으로 바인딩 되는거 아닐까요?

마르코가 말씀해주셨던 매개변수에서 event를 추가했을때 오류가 났던 부분을 조금 더 보면, bind 메서드를 이용해서 명시적으로 event 를 넘겨줄 수도 있긴한 것 같은데, e 라는 이름말고 window.event 라는 인자로 넘겨보면 작동되는걸 확인할 수 있을거에요. 그 결과는 해당 콜백함수에서 해당 이벤트가 발생하는 곳이 아닌 전역 객체의 이벤트를 인자로 받고있는걸 확인할 수 있어요! (CodeSandbox 환경에서 실험해봤는데 event 라는 이름으로 전역 이벤트를 넘겨줄 수도있긴한데 이건 쓰지말라네요.. 아마 이벤트를 바꾸는 경우는 많이 없지않을까요?)

call 메서드는 즉시 실행함수라서 해당 메서드를 이용해서 콜백함수를 만든다는 생각은 안해봤는데 화살표 함수 속에 call 메서드를 넣어주는 방식으로 사용해보셨네요!
그런데 화살표 함수의 특징이 실행 컨텍스트 생성 과정에서 this 바인딩이 제외되고 this가 상위 컨텍스트를 가리킨다는 특징이 있어서, call 메서드로 따로 this를 바인딩 안해줘도 this가 가리키는 대상이 같지않을까 싶어요! 
다만 this 바인딩 대상이 상위컨텍스트가 아닌 원하는 this가 따로 있다면 그때는 더 유용한 방법이겠네요.

마르코 질문 덕분에 이것저것 실험해보고 한층 더 생각해볼 수 있었네요 감사합니다👏

**Reply by onschan** [reply]
조금 더 생각해보자면, 마르코가 말씀해주신 세 가지 방법 모두 removeEventListener 의 콜백함수로는 다시 넘겨주기 힘들거에요! 그러면 어떻게 removeEventListener 로 같은 콜백함수를 다시 넘겨줘서 이벤트 등록 해제를 할 수 있는지 생각해봐도 좋을거같아요!

**Reply by wonsss** [reply]
아 그리고 저는 몰랐는데, 방금 책을 읽다보니까 왜 없던 e가 넘어가는지 알게 됐어요.

- 콜백함수의 인자에 대한 제어권을 addEventListener가 가진 상태이며, addEventListener는 콜백 함수를 호출할 때 첫 번째 인자에 `이벤트 객체`를 주입한다.

**Comment by wonsss** [comment]
> bind 메서드 사용에서는 함수 실행이 아닌 새로 생성된 함수 자체가 콜백함수로 넘어갔기 때문에 이벤트 인자가 자동으로 바인딩 되는거 아닐까요?

아하 그럴 수 있겠네요! 

> (call 메서드 관련) 다만 this 바인딩 대상이 상위컨텍스트가 아닌 원하는 this가 따로 있다면 그때는 더 유용한 방법이겠네요.

그렇겠네요. 사실 해당 화살표함수 사용 방식에서는 call(this)는 사족이긴 한데, 상황에 따라 유용한 경우도 있겠네요.

>  어떻게 removeEventListener 로 같은 콜백함수를 다시 넘겨줘서 이벤트 등록 해제를 할 수 있는지 생각해봐도 좋을거같아요!

아주 좋은 고민거리인 것 같아서, 이 부분은 [별도의 디스커션](https://github.com/woowacourse-study/2022-book-muscle/discussions/50)에서 진행하면 좋을 것 같아요.

감사합니다. 온스타 😄
